proposals/0000-decl-invis-binders.rst

```diff
@@ -0,0 +1,621 @@
+      | otherwise     = error "Unzippable quantifier/binder pair"
+    zipSAKS [] (b:bs) = error "Too many binders"
+    zipSAKS _ [] = []
+
+    skippable q = isInvisibleQuantifier q
+    zippable q b =
+      (isInvisibleForall q && isInvisibleBinder b) ||
+      (isVisibleQuantifier q && isVisibleBinder b)
+
+   ``isInvisibleForall`` holds for the ``forall x.`` quantifier only, but does
+   not hold for ``forall {x}.`` or any other quantifier.
+
+   ``isInvisibleBinder`` holds for ``@k``, ``@(k :: s)``, and ``@_``, but does
+   not hold for ``k`` or ``(k :: s)``. ``isVisibleBinder`` is the opposite.
+
+   ``isVisibleQuantifier`` holds for ``a ->`` and ``forall x ->`` only.
+   ``isInvisibleQuantifier`` is the opposite.
+
+   This implies that within a single declaration, some ``forall x.``
+   quantifiers may be zipped with binders, while others are not.
+
+Secondary Change: Arity Inference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+4. During arity inference, do not include trailing ``forall x.`` quantifiers in
+   the arity.
+
+Secondary Change: Scoping Rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+5. In type synonym declarations, require that every variable mentioned on the
+   RHS must be bound on the LHS.
+
+6. In type family and data family instances, require that every variable
+   mentioned on the RHS must also occur on the LHS.
+
+Secondary Change: Instantiation Rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+7. In type family and data family instances, the instantiation is fully
+   determined by the left hand side, without looking at the right hand side.
+
+Examples
+--------
+
+Here's an example from the wild (thanks Jakob Bruenker)::
+
+  data Relation n m = MkR
+
+  type Trans :: forall a b c . Relation a b -> Relation b c -> Relation a c
+  type family Trans pa pb where
+    Trans rel MkR = rel -- this type checks but is a partial type family
+
+In current GHC this typechecks, but the type family is not total. Why? Because
+the fully-explicit version is::
+
+  type family Trans pa pb where
+    Trans @a @b @b rel MkR = rel
+
+Notice the repeated ``b`` on the LHS. The author was entirely unaware that the
+resulting type family was partial, because the equation he wrote looked total.
+With change (7), the original program::
+
+  type family Trans pa pb where
+    Trans rel MkR = rel
+
+would be rejected. Why? Because the LHS imposes no kind constraints, so we get::
+
+  type family Trans pa pb where
+    Trans @a @b @c (rel :: Relation a b) (MkR :: Relation b c) = ...
+
+so the RHS must have kind ``Relation a c``. But it doesn't; ``rel`` has
+kind ``Relation a b``. So the declaration is rejected, which would have
+saved Jakob some time.
+
+Effect and Interactions
+-----------------------
+
+* Changes (1), (2), and (3) provide the programmer with a more principled way
+  of brining type variables into scope in certain corner cases.
+
+* Changes (4), (5), and (6) simplify arity inference and scoping rules, but they are
```

Comment:
(6) does not need a migration strategy, because the fix is backward-compatible.

I'm not very bothered about (4), which is rather exotic.

For (5), I would say to issue a warning for a few releases before pulling the trigger. I don't think that would be hard to implement.